contracts/smart-account/modules/SessionValidationModules/ABISessionValidationModule.sol

```diff
@@ -0,0 +1,164 @@
+                bytes4(_op.callData[0:4]) == EXECUTE_SELECTOR,
+            "ABISV Invalid Selector"
+        );
+
+        // we expect _op.callData to be `SmartAccount.execute(to, value, calldata)` calldata
+        (address destinationContract, uint256 callValue, ) = abi.decode(
+            _op.callData[4:], // skip selector
+            (address, uint256, bytes)
+        );
+
+        bytes calldata data;
+        {
+            //offset represents where does the inner bytes array start
+            uint256 offset = uint256(bytes32(_op.callData[4 + 64:4 + 96]));
+            uint256 length = uint256(
+                bytes32(_op.callData[4 + offset:4 + offset + 32])
+            );
+            data = _op.callData[4 + offset + 32:4 + offset + 32 + length];
+        }
+
+        return
+            validateSessionParams(
+                destinationContract,
+                callValue,
+                data,
+                _sessionKeyData,
+                new bytes(0)
+            ) ==
+            ECDSA.recover(
+                ECDSA.toEthSignedMessageHash(_userOpHash),
+                _sessionKeySignature
+            );
+    }
+
+    /**
+     * @dev validates that the call (destinationContract, callValue, funcCallData)
+     * complies with the Session Key permissions represented by sessionKeyData
+     * @param destinationContract address of the contract to be called
+     * @param callValue value to be sent with the call
+     * @param _funcCallData the data for the call. is parsed inside the SVM
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * param _callSpecificData additional data, for example some proofs if the SVM utilizes merkle trees itself
+     * for example to store a list of allowed tokens or receivers
+     */
+    function validateSessionParams(
+        address destinationContract,
+        uint256 callValue,
+        bytes calldata _funcCallData,
+        bytes calldata _sessionKeyData,
+        bytes memory /*_callSpecificData*/
+    ) public pure virtual override returns (address) {
+        (address sessionKey, Permission memory permission) = abi.decode(
+            _sessionKeyData,
+            (address, Permission)
+        );
+
+        if (destinationContract != permission.destinationContract) {
+            revert("ABISV Wrong Destination");
+        }
+        if (callValue > permission.valueLimit) {
+            revert("ABISV Value exceeded");
+        }
+
+        if (!checkPermission(_funcCallData, permission))
+            revert("ABISV: Permission violated");
+
+        return sessionKey;
+    }
+
+    function checkPermission(
+        bytes calldata data,
+        Permission memory permission
+    ) internal pure returns (bool) {
+        if (bytes4(data[0:4]) != permission.selector) return false;
+        uint256 length = permission.rules.length;
+        for (uint256 i; i < length; ) {
+            Rule memory rule = permission.rules[i];
```

Comment:
`rule.offset`, `rule.condition` and `rule.value` can be cached on the stack to save some gas